Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp

```diff
@@ -258,284 +265,302 @@ constexpr int GetCharacterCount(MemoryViewWidget::Type type)
+    for (int c = 0; c < m_table->columnCount() - MISC_COLUMNS; c++)
+    {
+      auto* item = m_table->item(i, c + MISC_COLUMNS);
 
-    starting_column += column_count;
+      u32 item_address;
+      if (m_dual_view && c >= data_span)
+        item_address = row_address + (c - data_span) * GetTypeSize(m_type);
+      else
+        item_address = row_address + c * GetTypeSize(m_type);
+
+      item->setData(USER_ROLE_CELL_ADDRESS, item_address);
+    }
   }
 
-  UpdateColumns(m_type, starting_column);
+  UpdateColumns();
   UpdateBreakpointTags();
 
-  m_table->setColumnWidth(0, m_table->rowHeight(0));
-
-  for (int i = starting_column; i <= m_table->columnCount(); i++)
-    m_table->setColumnWidth(i, m_font_width * GetCharacterCount(m_type));
-
   m_table->viewport()->update();
   m_table->update();
   update();
 }
 
-void MemoryViewWidget::UpdateColumns(Type type, int first_column)
+void MemoryViewWidget::UpdateColumns()
 {
-  if (Core::GetState() != Core::State::Paused)
+  // Check if table is created
+  if (m_table->item(1, 1) == nullptr)
     return;
 
-  const int data_columns = m_bytes_per_row / GetTypeSize(type);
-  const AddressSpace::Accessors* accessors = AddressSpace::GetAccessors(m_address_space);
-
-  auto text_alignment = Qt::AlignLeft;
-  if (type == Type::Signed32 || type == Type::Unsigned32 || type == Type::Signed16 ||
-      type == Type::Unsigned16 || type == Type::Signed8 || type == Type::Unsigned8)
-  {
-    text_alignment = Qt::AlignRight;
-  }
+  const QSignalBlocker blocker(m_table);
 
   for (int i = 0; i < m_table->rowCount(); i++)
   {
-    u32 row_address = m_table->item(i, 1)->data(USER_ROLE_CELL_ADDRESS).toUInt();
-    if (!accessors->IsValidAddress(row_address))
-      continue;
+    for (int c = MISC_COLUMNS; c < m_table->columnCount(); c++)
+    {
+      auto* cell_item = m_table->item(i, c);
+      const u32 cell_address = cell_item->data(USER_ROLE_CELL_ADDRESS).toUInt();
+      const Type type = static_cast<Type>(cell_item->data(USER_ROLE_VALUE_TYPE).toInt());
 
-    auto update_values = [&](auto value_to_string) {
-      for (int c = 0; c < data_columns; c++)
-      {
-        auto* cell_item = new QTableWidgetItem;
-        cell_item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsSelectable | Qt::ItemIsEditable);
-        cell_item->setTextAlignment(text_alignment);
+      cell_item->setText(ValueToString(cell_address, type));
+    }
+  }
+}
 
-        const u32 cell_address = row_address + c * GetTypeSize(type);
+QString MemoryViewWidget::ValueToString(u32 address, Type type)
+{
+  const AddressSpace::Accessors* accessors = AddressSpace::GetAccessors(m_address_space);
+  if (!accessors->IsValidAddress(address) || Core::GetState() != Core::State::Paused)
+    return QStringLiteral("-");
+
+  switch (type)
+  {
+  case Type::Hex8:
+  {
+    const u8 value = accessors->ReadU8(address);
+    return QStringLiteral("%1").arg(value, 2, 16, QLatin1Char('0'));
+    break;
+  }
```

Comment:
The `break` statement can be removed because it's unreachable after a `return` statement. Ditto for all the other cases.